File: internal/installer/plan.go
package installer

import (
	"context"
	"os"
	"os/exec"
	"strings"
	"time"
)

type Task struct {
	ComponentID  string
	Label        string
	Operation    string
	Dependencies []string
}

type Skip struct {
	ComponentID string
	Reason      string
}

type Result struct {
	Task     Task
	Status   string
	Output   string
	Started  time.Time
	Finished time.Time
}

type Environment struct {
	Commands map[string]bool
}

func DetectEnvironment() Environment {
	commands := make(map[string]bool)
	for _, name := range []string{"brew", "xcode-select", "git", "gh", "code", "php", "composer", "opencode"} {
		_, err := exec.LookPath(name)
		commands[name] = err == nil
	}
	return Environment{Commands: commands}
}

type Runner func(context.Context, string) (string, error)

func ShellRunner(ctx context.Context, command string) (string, error) {
	cmd := exec.CommandContext(ctx, "sh", "-c", command)
	cmd.Env = append(os.Environ(), "HOMEBREW_NO_AUTO_UPDATE=1", "HOMEBREW_NO_ENV_HINTS=1")
	output, err := cmd.CombinedOutput()
	return string(output), err
}

func Plan(profile Profile, env Environment) ([]Task, []Skip, error) {
	return PlanWithApplied(profile, env, nil)
}

func PlanWithApplied(profile Profile, env Environment, applied map[string]bool) ([]Task, []Skip, error) {
	ordered := make([]Component, 0, len(Components()))
	visiting, visited := map[string]bool{}, map[string]bool{}
	var add func(Component)
	add = func(component Component) {
		if visited[component.ID] || visiting[component.ID] {
			return
		}
		visiting[component.ID] = true
		for _, dependency := range component.Dependencies {
			for _, candidate := range Components() {
				if candidate.ID == dependency {
					add(candidate)
				}
			}
		}
		visiting[component.ID] = false
		visited[component.ID] = true
		ordered = append(ordered, component)
	}
	for _, component := range Components() {
		if profile.Components[component.ID] && !applied[component.ID] {
			add(component)
		}
	}

	var tasks []Task
	var skips []Skip
	for _, component := range ordered {
		for _, command := range component.Commands {
			if command == "xcode-select --install" && env.Commands["xcode-select"] {
				continue
			}
			if strings.HasPrefix(command, "brew ") && !env.Commands["brew"] {
				skips = append(skips, Skip{ComponentID: component.ID, Reason: "Homebrew is not installed"})
				break
			}
			tasks = append(tasks, Task{ComponentID: component.ID, Label: component.Label, Operation: command, Dependencies: component.Dependencies})
		}
	}
	return tasks, skips, nil
}

type Progress func(Task)

func Execute(ctx context.Context, tasks []Task, run Runner) []Result {
	return ExecuteWithProgress(ctx, tasks, run, nil)
}

func ExecuteWithProgress(ctx context.Context, tasks []Task, run Runner, progress Progress) []Result {
	results := make([]Result, 0, len(tasks))
	failed := map[string]bool{}
	for _, task := range tasks {
		started := time.Now()
		blocked := false
		for _, dependency := range task.Dependencies {
			if failed[dependency] {
				blocked = true
				break
			}
		}
		if blocked {
			results = append(results, Result{Task: task, Status: "skipped", Output: "dependency failed", Started: started, Finished: time.Now()})
			continue
		}
		if ctx.Err() != nil {
			results = append(results, Result{Task: task, Status: "skipped", Output: "cancelled", Started: started, Finished: time.Now()})
			continue
		}
		if progress != nil {
			progress(task)
		}
		output, err := run(ctx, task.Operation)
		status := "installed"
		if err != nil {
			status = "failed"
			failed[task.ComponentID] = true
		}
		results = append(results, Result{Task: task, Status: status, Output: output, Started: started, Finished: time.Now()})
	}
	return results
}

File: tools/tui/src/plan.ts
import type { ContextPackage } from "./context";

// Types kept from the merged executor (installer-execute spec): a Task is one
// operation; statuses are the exact strings the specs pin.
export interface Task {
  componentId: string;
  label: string;
  operation: string;
  dependencies: string[];
}

export interface Result {
  task: Task;
  status: "installed" | "failed" | "skipped";
  output: string;
  started: Date;
  finished: Date;
}

export type Runner = (
  operation: string,
  signal?: AbortSignal,
) => Promise<{ output: string; err?: Error }>;

export type Progress = (task: Task) => void;

/**
 * One package row -> its brew command, or null for delegated topic rows.
 */
export function brewCommandFor(packageRow: ContextPackage): string | null {
  if (packageRow.kind === "tap") return `brew tap ${packageRow.id}`;
  if (packageRow.kind === "brew") return `brew install ${packageRow.id}`;
  if (packageRow.kind === "cask") return `brew install --cask ${packageRow.id}`;
  return null; // kind "topic" is delegated to `dot install`, never planned here.
}

// Production runner ported from ShellRunner: sh -c with Homebrew update/hint
// suppression inherited on top of the current environment, combined stream
// capture. A non-zero exit maps to err while output is preserved.
export const shellRunner: Runner = async (operation, signal) => {
  const proc = Bun.spawn(["sh", "-c", operation], {
    env: {
      ...process.env,
      HOMEBREW_NO_AUTO_UPDATE: "1",
      HOMEBREW_NO_ENV_HINTS: "1",
    },
    stdout: "pipe",
    stderr: "pipe",
    signal,
  });
  const [stdout, stderr] = await Promise.all([
    new Response(proc.stdout).text(),
    new Response(proc.stderr).text(),
  ]);
  const output = stdout + stderr;
  const exitCode = await proc.exited;
  if (exitCode === 0) {
    return { output };
  }
  return { output, err: new Error(`command exited with code ${exitCode}`) };
};

// Sequential executor ported from ExecuteWithProgress. Check order matches the
// Go original exactly: blocked dependencies first, then cancellation, then
// progress fires immediately before the runner is invoked.
export async function executeWithProgress(
  tasks: Task[],
  run: Runner,
  signal?: AbortSignal,
  progress?: Progress,
  isCancelled?: (task: Task) => boolean,
): Promise<Result[]> {
  const results: Result[] = [];
  const failed = new Set<string>();
  for (const task of tasks) {
    const started = new Date();
    let blocked = false;
    for (const dependency of task.dependencies) {
      if (failed.has(dependency)) {
        blocked = true;
        break;
      }
    }
    if (blocked) {
      results.push({
        task,
        status: "skipped",
        output: "dependency failed",
        started,
        finished: new Date(),
      });
      continue;
    }
    if (signal?.aborted || isCancelled?.(task)) {
      results.push({
        task,
        status: "skipped",
        output: isCancelled?.(task) ? "interrupted" : "cancelled",
        started,
        finished: new Date(),
      });
      continue;
    }
    if (progress) {
      progress(task);
    }
    const { output, err } = await run(task.operation, signal);
    let status: Result["status"] = "installed";
    if (err) {
      status = "failed";
      failed.add(task.componentId);
    }
    results.push({ task, status, output, started, finished: new Date() });
  }
  return results;
}
